### ISSUE TITLE
AV-huBERT as backbone for a slightly different task

### ISSUE THREAD

**Issue by gadese-vooban** [facebookresearch/av_hubert#95]
Hi,

I've been looking into doing transfer learning with the model for a task similar to lip-reading, but slightly different.

For a given video with people speaking, I need to predict the mouth shape associated with different phonemes. For example, the "O" sound has an open mouth. The "F" sound has the upper teeth touching the lower lips, etc. My data includes small video clips (between 2 and 30 frames) with audio (the input) that are associated to a specific mouth shape (the label).

I figured I'd used a pretrained model like AV-HuBERT as the backbone for my task. The repo is very well structured and maintained, but I have to admit I get a bit lost in the code.

How would you approach the modification of your model for my task? The way I understand it, there are 2 possible approaches:

1. Using the model as a frozen feature generator
- Feed my videos into the model
- Extract features from the last layer
- Use the features as the input to a new model/classifier and train this way
2. Traditional transfer learning
- Modify the model and scripts to accomodate my new learning task
- Train for the new task with the pretrained checkpoints as a starting point.

Approach 2 seems like it would be better, but tbh I'm not sure where to start to modify the model/scripts.

Any insight? Thanks!

**Comment by MiscellaneousStuff** [comment]
Hello,

A few months ago I worked on a [project](https://github.com/MiscellaneousStuff/comp-vis-avhubert/blob/main/main.ipynb) which used AV-HuBERT as a frozen feature extractor like you described in approach 1. I then trained a simple MLP classifier layer on top of it to predict which phoneme was being uttered by a person, per timestep (so depending on which video frame) in a video and it achieved around 65% accuracy. If you're also interested which phonemes are being classified well, and which phonemes are often being confused, this [paper](https://www.joemakepeace.co.uk/lip2seq.pdf) provides useful graphs (refer to the larger dataset) which you may find useful and will let you have realistic expectations about performance (hopefully).

The code for that is in the notebook which I attached and you can basically re-use the code which I provided in that notebook but just change your target labels. I think you should also roughly expect around the same accuracy seeing as predicting the phoneme is quite similar to predicting the mouth shape during the phonemes utterance?

Re-tasking the AV-HuBERT model to re-train for your new task would be a lot more effort to be honest and it also removes the benefit which you already have where you can just pre-compute all of your datasets AV-HuBERT features and very quickly just train the mouth shape classifier using the pre-computed features. I found that in my project as I was only training a very simple MLP classifier on top of the AV-HuBERT features, it was extremely quick to train the model which is also useful for experimentation speed (you can find out what is going to work a lot faster than trying approach 2 which is also super useful for you).

Good luck!

**Comment by gadese-vooban** [comment]
> Hello,
> 
> A few months ago I worked on a [project](https://github.com/MiscellaneousStuff/comp-vis-avhubert/blob/main/main.ipynb) which used AV-HuBERT as a frozen feature extractor like you described in approach 1. I then trained a simple MLP classifier layer on top of it to predict which phoneme was being uttered by a person, per timestep (so depending on which video frame) in a video and it achieved around 65% accuracy. If you're also interested which phonemes are being classified well, and which phonemes are often being confused, this [paper](https://www.joemakepeace.co.uk/lip2seq.pdf) provides useful graphs (refer to the larger dataset) which you may find useful and will let you have realistic expectations about performance (hopefully).
> 
> The code for that is in the notebook which I attached and you can basically re-use the code which I provided in that notebook but just change your target labels. I think you should also roughly expect around the same accuracy seeing as predicting the phoneme is quite similar to predicting the mouth shape during the phonemes utterance?
> 
> Re-tasking the AV-HuBERT model to re-train for your new task would be a lot more effort to be honest and it also removes the benefit which you already have where you can just pre-compute all of your datasets AV-HuBERT features and very quickly just train the mouth shape classifier using the pre-computed features. I found that in my project as I was only training a very simple MLP classifier on top of the AV-HuBERT features, it was extremely quick to train the model which is also useful for experimentation speed (you can find out what is going to work a lot faster than trying approach 2 which is also super useful for you).
> 
> Good luck!

Glad to know I'm not the only one looking into similar projects! I'll have a look at your code, thanks :)

Did you use the training data in the lip2seq paper for your task? Or did you have your own data?

**Comment by MiscellaneousStuff** [comment]
I used the Jordan Peterson lectures on YouTube as I needed many hours of a single speaker speaking for long periods of time with their lips being clearly visible. However from what I've seen, most spoken datasets will have a similar phoneme distribution to what is reported in the paper (which should also be very relevant to your project as the phonemes and mouth movements will be closely related).
